Qutrit channel operation

Tarun-Kumar07 opened the pull request:
**Context:**
The goal of this PR is to create a qutrit counterpart to the QubitChannel, enabling noise to be defined using a set of 
3×3 Kraus operators.

**Description of the Change:**
- Implement `QutritChannel` similar to `QubitChannel`

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
Fixes #5649

mudit2812 commented on pennylane/ops/qutrit/channel.py:
```diff
@@ -313,3 +313,74 @@ def compute_kraus_matrices(gamma_1, gamma_2):  # pylint:disable=arguments-differ
+                raise ValueError("Only trace preserving channels can be applied.")
+
+    def _flatten(self):
+        return (self.data,), (self.wires, ())
```
You can test it similar to [this test](https://github.com/PennyLaneAI/pennylane/blob/7384db3693cef7d04d2414e021714791a11b0d42/tests/test_operation.py#L326).

mudit2812 commented on doc/releases/changelog-dev.md:
```diff
@@ -151,6 +151,8 @@
 * ``qml.QutritDepolarizingChannel`` has been added, allowing for depolarizing noise to be simulated on the `default.qutrit.mixed` device.
   [(#5502)](https://github.com/PennyLaneAI/pennylane/pull/5502)
+ 
+* `qml.QutritChannel` has been added, enabling the specification of noise using a collection of (3x3) Kraus matrices on the `default.qutrit.mixed` device. [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5649)
```
```suggestion
* `qml.QutritChannel` has been added, enabling the specification of noise using a collection of (3x3) Kraus matrices on the `default.qutrit.mixed` device.
  [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5649)
```

mudit2812 commented on pennylane/ops/qutrit/channel.py:
```diff
@@ -340,3 +340,74 @@ def compute_kraus_matrices(gamma_1, gamma_2, gamma_3):  # pylint:disable=argumen
+
+class QutritChannel(Channel):
+    r"""
+    Apply an arbitrary fixed quantum channel.
```
```suggestion
    Apply an arbitrary fixed qutrit channel.
```

mudit2812 commented on pennylane/ops/qutrit/channel.py:
```diff
@@ -340,3 +340,74 @@ def compute_kraus_matrices(gamma_1, gamma_2, gamma_3):  # pylint:disable=argumen
+            raise ValueError("All Kraus matrices must have the same shape.")
+
+        # check the dimension of all Kraus matrices are valid
+        if any(K.shape[0] != QUDIT_DIM for K in K_list):
```
This condition will raise an error if there are multiple wires since the expected shape is `(3 ** num_wires, 3 ** num_wires)`.

```suggestion
        if any(K.shape[0] != QUDIT_DIM ** (len(self.wires)) for K in K_list):
```

mudit2812 reviewed (commented):
Thanks for opening the PR @Tarun-Kumar07 ! I've just left a couple of preliminary comments. I noticed that there aren't any tests for `QutritChannel`s with multiple wires. Could you include those as well? Additionally, I noticed that the tests call `channel.QutritChannel`, but this should be `qml.QutritChannel`.

Feel free to re-request review whenever the comments are resolved 😄

Tarun-Kumar07 commented:
Hi @mudit2812,

Thank you for the review. I have modified the unit test `test_kraus_matrices_are_dimensions` to ensure it runs on multiple wires. Additionally, I have adjusted the error message for clarity. Could you please review it again?

Thanks!

trbromley commented on doc/releases/changelog-dev.md:
```diff
@@ -179,6 +179,9 @@
   [(#5502)](https://github.com/PennyLaneAI/pennylane/pull/5502)
+ 
+* `qml.QutritChannel` has been added, enabling the specification of noise using a collection of (3x3) Kraus matrices on the `default.qutrit.mixed` device.
+  [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5649)
```
```suggestion
  [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5793)
```

mudit2812 commented on pennylane/ops/qutrit/channel.py:
```diff
@@ -378,10 +378,9 @@ def __init__(self, K_list, wires=None, id=None):
-            )
+        kraus_dim = QUDIT_DIM ** len(self.wires)
+        if any(K.shape[0] != kraus_dim for K in K_list):
+            raise ValueError(f"Dimension of all Kraus matrices must be ({kraus_dim},{kraus_dim}).")
```
```suggestion
            raise ValueError(f"Shape of all Kraus matrices must be ({kraus_dim},{kraus_dim}).")
```

mudit2812 reviewed (commented):
Looks great! I'll approve once CI is green. Nice work @Tarun-Kumar07

Tarun-Kumar07 commented:
Hi @mudit2812 and @trbromley , I am not sure why the test `upload-to-codecov` failed. Could you please help.
I have addressed rest of the code review comments, please let me know if something needs to improved.

Thanks !!

mudit2812 reviewed (changes requested):
One final change I forgot to request is to add `QutritChannel` to the operations page in the pennylane docs. [This](https://docs.pennylane.ai/en/latest/introduction/operations.html#qutrit-noisy-channels) is the section to add the operator, and [this](https://github.com/PennyLaneAI/pennylane/blob/38b3e74e26ef7f4f758e1cb34d4db0a1fc679780/doc/introduction/operations.rst?plain=1#L510) is the file that needs to be updated to add the operation.

mudit2812 reviewed (approved):
Nice work @Tarun-Kumar07 ! I'll pass this on to another team member now for a second review.

albi3ro commented on tests/ops/qutrit/test_qutrit_channel_ops.py:
```diff
@@ -288,3 +288,98 @@ def test_kraus_jac_jax(self):
+        ]
+
+        qutrit_channel = qml.QutritChannel(kraus, 1, id="test")
+        data, metadata = qutrit_channel._flatten()  # pylint: disable=protected-access
```
```suggestion
        data, metadata = qutrit_channel._flatten()  # pylint: disable=protected-access
        new_op = qml.QutritChannel._unflatten(data, metadata)
        assert qml.equal(qutrit_channel, new_op)
```

We should probably make sure the flatten method is compatible with the analogous `unflatten` method.

albi3ro commented on pennylane/ops/qutrit/channel.py:
```diff
@@ -457,3 +457,73 @@ def compute_kraus_matrices(p_01, p_02, p_12):  # pylint:disable=arguments-differ
+            if not math.allclose(Kraus_sum, math.eye(K_list[0].shape[0])):
+                raise ValueError("Only trace preserving channels can be applied.")
+
+    def _flatten(self):
```
For context, `Operator._flatten` and `Operator._unflatten` are used to improve out integration with jax, see [jax pytrees docs](https://jax.readthedocs.io/en/latest/pytrees.html).  Now jitted functions can can accept and return operators, and we can have improved parameter handling.

The contraints for these two methods are:

1) `type(op)._unflatten(*op._flatten())` should return the original operation.
2) Any traceable/ trainable data or any other pytrees should be placed in the `data` output
3) the `metadata` output should be hashable and immutable.

An example of this in action is:
```
op = qml.adjoint(2*qml.X(0)) @ qml.Rot(1.2, 2.3, 3.4, 0)
data, structure = qml.pytrees.flatten(op)
qml.pytrees.unflatten([3.0, -1.2, -2.3, -3.4], structure)
```

albi3ro replied:
Ignore me.  

This is due to the difference between `QutritChannel((matrix1, matrix2), wires)` instead of `QutritChannel(matrix1, matrix2, wires)`.  Makes sense now.

albi3ro reviewed (approved):
🎉

Tarun-Kumar07 commented:
Hi @albi3ro and @mudit2812 ,

Can this PR be merged as there are two approvals.

Thanks !!

Tarun-Kumar07 commented:
Hey @mudit2812 , I am not sure why `upload-to-codecov` is failing :).

mudit2812 commented:
> Hey @mudit2812 , I am not sure why `upload-to-codecov` is failing :).

I've retriggered the upload to codecov job. Sometimes it happens 😅

mudit2812 commented:
Not sure what's going on but it's failing repeatedly. I'll bring this up with the team.

mudit2812 commented:
@Tarun-Kumar07 for now it might be worth just running CI all over again by pushing an empty commit. You can do so using `git commit --allow-empty -m "Trigger CI"`
